[Woo POS] MVP order, cart, and miscellaneous events

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: https://github.com/woocommerce/woocommerce-ios/issues/13613
Closes: https://github.com/woocommerce/woocommerce-ios/issues/15018

## Description
This PR adds the POS track events discussed on pdfdoF-6hn-p2 for MVP. Final events and properties are still TBD, so most likely we'll need to update some before launch. Events are not validated yet, I'll do that when we have a final decision on all events and their properties, as its very time consuming.

You should see all events decorated with `pos_*` in the console, and as `woocommerceios_pos_*` in Track events.

I've opted to call analytics from ServiceLocator in most views. I'm sure we can inject some of them directly or through the viewModel but didn't seem worthy just for a single event or two. Happy to iterate on this as needed.

## Testing

Go through POS performing different actions like adding, removing items from cart, checkout, create order, pay by cash or card, send receipt, exit POS, get support, etc, ... I've tested all of them manually, and the ones where we inject analytics have unit tests as well, so feel free to go through all of them or just some.

- order_creation_success			
```
🔵 Tracked pos_order_creation_success, properties: [plan: , blog_id: -1, was_ecommerce_trial: false, is_wpcom_store: false, site_url: https://indiemelon.mystagingwebsite.com, store_id: c5bd46cc-1804-4f7b-badb-bb98c449127f]
```

- order_creation_failed			
```
🔵 Tracked pos_order_creation_failed, properties: [plan: , store_id: c5bd46cc-1804-4f7b-badb-bb98c449127f, was_ecommerce_trial: false, site_url: https://indiemelon.mystagingwebsite.com, is_wpcom_store: false, blog_id: -1, error_context: updateOrderFailed, use_gift_card: false, error_description: The operation couldn’t be completed. (Yosemite.POSOrderService.(unknown context at $10975b684).POSOrderServiceError error 1.)]
```

- create_new_order_tapped 			
```
🔵 Tracked pos_create_new_order_tapped, properties: [blog_id: -1, store_id: c5bd46cc-1804-4f7b-badb-bb98c449127f, was_ecommerce_trial: false, site_url: https://indiemelon.mystagingwebsite.com, plan: , is_wpcom_store: false]
```

- item_removed_from_cart			
```
🔵 Tracked pos_item_removed_from_cart, properties: [plan: , was_ecommerce_trial: false, site_url: https://indiemelon.mystagingwebsite.com, store_id: c5bd46cc-1804-4f7b-badb-bb98c449127f, blog_id: -1, is_wpcom_store: false]
```

- checkout_tapped (prop: items_in_cart with the number of items in cart when checkout is tapped)			
```
🔵 Tracked pos_checkout_tapped, properties: [is_wpcom_store: false, blog_id: -1, items_in_cart: 1, store_id: c5bd46cc-1804-4f7b-badb-bb98c449127f, was_ecommerce_trial: false, site_url: https://indiemelon.mystagingwebsite.com, plan: ]
```

- back_to_cart_tapped			
```
🔵 Tracked pos_back_to_cart_tapped, properties: [is_wpcom_store: false, blog_id: -1, plan: , was_ecommerce_trial: false, store_id: c5bd46cc-1804-4f7b-badb-bb98c449127f, site_url: https://indiemelon.mystagingwebsite.com]
```

- get_support_tapped			
```
🔵 Tracked pos_get_support_tapped, properties: [plan: , blog_id: -1, was_ecommerce_trial: false, is_wpcom_store: false, site_url: https://indiemelon.mystagingwebsite.com, store_id: c5bd46cc-1804-4f7b-badb-bb98c449127f]
```

- exit_pos_menu_item_tapped			
```
🔵 Tracked pos_exit_pos_menu_item_tapped, properties: [blog_id: -1, site_url: https://indiemelon.mystagingwebsite.com, plan: , was_ecommerce_trial: false, is_wpcom_store: false, store_id: c5bd46cc-1804-4f7b-badb-bb98c449127f]
```

- exit_pos_confirmed			
```
🔵 Tracked pos_exit_pos_confirmed, properties: [site_url: https://indiemelon.mystagingwebsite.com, store_id: c5bd46cc-1804-4f7b-badb-bb98c449127f, plan: , is_wpcom_store: false, was_ecommerce_trial: false, blog_id: -1]
```

- simple_products_explanation_dialog_shown			
```
🔵 Tracked pos_simple_products_explanation_dialog_shown, properties: [plan: , store_id: c5bd46cc-1804-4f7b-badb-bb98c449127f, blog_id: -1, is_wpcom_store: false, site_url: https://indiemelon.mystagingwebsite.com, was_ecommerce_trial: false]
```

- email_receipt_tapped			
```
🔵 Tracked pos_email_receipt_tapped, properties: [blog_id: -1, site_url: https://indiemelon.mystagingwebsite.com, is_wpcom_store: false, plan: , was_ecommerce_trial: false, store_id: c5bd46cc-1804-4f7b-badb-bb98c449127f]
```

- email_receipt_send_tapped			
```
🔵 Tracked pos_email_receipt_send_tapped, properties: [plan: , blog_id: -1, site_url: https://indiemelon.mystagingwebsite.com, is_wpcom_store: false, store_id: c5bd46cc-1804-4f7b-badb-bb98c449127f, was_ecommerce_trial: false]
```

- products_pull_to_refresh			
```
🔵 Tracked pos_products_pull_to_refresh, properties: [blog_id: -1, was_ecommerce_trial: false, store_id: c5bd46cc-1804-4f7b-badb-bb98c449127f, is_wpcom_store: false, plan: , site_url: https://indiemelon.mystagingwebsite.com]
```

- variations_pull_to_refresh			
```
🔵 Tracked pos_variations_pull_to_refresh, properties: [site_url: https://indiemelon.mystagingwebsite.com, store_id: c5bd46cc-1804-4f7b-badb-bb98c449127f, is_wpcom_store: false, was_ecommerce_trial: false, blog_id: -1, plan: ]
```

- pos_loaded (prop milliseconds_time_elapsed_in_splash_screen with the milliseconds it took between view appear and disappear )			
```
🔵 Tracked pos_pos_loaded, properties: [was_ecommerce_trial: false, is_wpcom_store: false, store_id: c5bd46cc-1804-4f7b-badb-bb98c449127f, site_url: https://indiemelon.mystagingwebsite.com, plan: , milliseconds_time_elapsed_in_splash_screen: 1995.630126953125, blog_id: -1]
```

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

iamgabrielma committed:
track success-failure event for new POS orders

iamgabrielma committed:
track POS success view event flows for new order and receipts

iamgabrielma committed:
add item_removed_from_cart POS event

iamgabrielma committed:
track POS events checkout and back to edit cart

iamgabrielma committed:
track POS exit attempt and confirmation

dangermattic commented:
<!--
  0 Errors
  2 Warnings: View files have been modified,...
  1 Message: <p>This PR contains changes to...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/237">21.7</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false"><p>This PR contains changes to Tracks-related logic. Please ensure (<strong>author and reviewer</strong>) the following are completed:</p>

<ul>
  <li>The tracks events must be validated in the Tracks system.</li>
  <li>Verify the internal Tracks spreadsheet has also been updated.</li>
  <li>Please consider registering any new events.</li>
  <li>The PR must be assigned the <strong>category: tracks</strong> label.</li>
</ul></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
POS get support tapped event

iamgabrielma committed:
POS simple products dialog event

iamgabrielma committed:
POS track pull to refresh products and variations

iamgabrielma committed:
Update checkout event with items in cart

iamgabrielma committed:
add pos_loaded event and milliseconds prop

iamgabrielma committed:
Add support request viewed to exempted from POS

We already track this event by checking the tap on the support button via “pos_get_support_tapped”. But “support_new_request_viewed” will get decorated with “pos_” as we do not exit POS mode to send the support request.

iamgabrielma committed:
Add pos events tests

Declaring WooAnalytics as dependency at file level makes the test suite crash when ran as a whole but not independently test by test. This is resolved by injecting the dependency at the test level, most likely a threading issue?

iamgabrielma commented on WooCommerce/Classes/Analytics/WooAnalyticsStat.swift:
```diff
@@ -1268,7 +1268,20 @@ enum WooAnalyticsStat: String {
     case backgroundUpdatesDisabled = "background_updates_disabled"
 
     // MARK: Point of Sale events
+    case pointOfSaleLoaded = "pos_loaded"
```
This one and others that contain "_pos_" in the event itself most likely will change, as "_pos_" will be repeated otherwise. Asked here: pdfdoF-6hn#comment-7484-p2

jaclync replied:
Personally, I'd prefer having an explicit `pos_` name here for at least POS-only events without the automatic prefix - I was a bit confused reviewing this PR first, not knowing the automatic prefix before.

In any case, as we are not changing the current implementation, I agree on removing the redundant `pos` in the main event name.

iamgabrielma replied:
Sorry I didn't thought about the AFK, I could have added a link to the previous PR for context 🙇 . I agree it's something I'd like to see explicitly as well as makes much easier to reason about.

iamgabrielma commented on WooCommerce/Classes/POS/Controllers/PointOfSaleOrderController.swift:
```diff
@@ -60,16 +64,24 @@ final class PointOfSaleOrderController: PointOfSaleOrderControllerProtocol {
+            }
         } catch {
-            DDLogError("🔴 [POS] Error syncing order: \(error)")
+            if isNewOrder {
```
I recall a conversation where we'll be moving from the current approach of updating/creating an order to always creating it when we go back, but I couldn't find it. For now this will log the event only when creating new orders

iamgabrielma commented on WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/PointOfSaleLoadingView.swift:
```diff
@@ -13,6 +15,25 @@ struct PointOfSaleLoadingView: View {
+
+    func trackElapsedTimeOnDisappear() {
+        if let appearTimeInMilliseconds = appearTimeInMilliseconds {
+            let elapsedTimeInMilliseconds = (Date().timeIntervalSince1970 * Constants.toMilliseconds) - appearTimeInMilliseconds
```
Used `.now()` and `.uptimeNanoseconds` then divide by 1_000_000 initially, but AI says that TimeInterval + multiplying it by 1000 is closer to Android's implementation. I have no special preference 🤔.

jaclync replied:
💬 It looks like we have a helper `WaitingTimeTracker` to track the elapsed time. How about using it with a new scenario?

iamgabrielma replied:
Good to know! Used here: 2f235538944ef962e94fdc697d2f098b93b91232

iamgabrielma commented on WooCommerce/WooCommerceTests/POS/Controllers/PointOfSaleOrderControllerTests.swift:
```diff
@@ -27,12 +29,19 @@ struct PointOfSaleOrderControllerTests {
     }
 
     @Test func syncOrder_with_cart_matching_order_doesnt_call_orderService() async throws {
+        let mockAnalyticsProvider = MockAnalyticsProvider()
```
Declaring analytics on the test level and running the suite seems to crash the tests. This doesn't happen if ran individually, just as a whole. 
```
WooCommerce(27810,0x16e0e3000) malloc: *** error for object 0x600000d64450: pointer being freed was not allocated
WooCommerce(27810,0x16e0e3000) malloc: *** set a breakpoint in malloc_error_break to debug
```

Declaring and injecting them in the SUT at the test level works well. I'm unsure of the issue but seems to indicate memory corruption in the analytics internals.

jaclync replied:
> Declaring analytics on the test level and running the suite seems to crash the tests.

Did you mean making some changes in this test caused the tests to crash? I tried commenting out these diffs in the test and running the test suite as a whole but I only saw crashes/errors from other test cases (I commented out these other tests and the test suite passed).

If we want to test the analytics events like in [L43 and L168](https://github.com/woocommerce/woocommerce-ios/pull/15047/files#diff-b2da4a624dc5cef2722f4ad046beed9819f9af13b47a1fe03a2424b1d27297aeL35), a separate test case would make the goal of the test more clear.

iamgabrielma replied:
> Did you mean making some changes in this test caused the tests to crash?

Injecting analytics dependencies at the top when we `init()` the struct makes these test crash indeed. I saw the same happening on a [different PR](https://github.com/woocommerce/woocommerce-ios/pull/15022#issuecomment-2627419452) when injecting `stores` at the suite level. 

My current understanding is that SwiftTesting runs these on a background thread by default, which doesn't play well with our action dispatcher having to run in the main thread. I suspected the same in this case when injecting analytics but the error message was different pointing to a memory issue.

The crash only happens when we run the whole suite, and not when tests are run independently. It's also resolved if we declare the dependencies scoped to the test level (as this PR), so it seems to indicate that is indeed a multi-threading issue?

iamgabrielma replied:
Actually, I think the issues I saw with crashing while testing were self-caused. I'm used to have a setup when using XCTest, but in the SwiftTesting context we have sub-structs inside the main testing suite, declaring the dependencies in those child-structs works without crashing Eg: `PointOfSaleAggregateModelTests.AnalyticsTests... `

iamgabrielma replied:
> If we want to test the analytics events like in [L43 and L168](https://github.com/woocommerce/woocommerce-ios/pull/15047/files#diff-b2da4a624dc5cef2722f4ad046beed9819f9af13b47a1fe03a2424b1d27297aeL35), a separate test case would make the goal of the test more clear.

True that, moved the order success/failure tests to their own tests on 2b362bc41b04a1841a939319aad2b12970e9b964

iamgabrielma commented on WooCommerce/Classes/Analytics/TracksProvider.swift:
```diff
@@ -116,6 +116,7 @@ private extension TracksProvider {
         let exemptedEvents: Set<String> = [
             "application_opened",
             "application_closed",
+            "support_new_request_viewed",
```
The support view opens in a modal within POS, since we do not exit POS mode this event gets decorated as well with the `woocommerceios_pos_` bit unless we explicitly filter it out.

jaclync replied:
I think an allowed list would be safer in case we didn't catch any events that shouldn't be decorated, also left a comment in https://github.com/woocommerce/woocommerce-ios/pull/15006#issuecomment-2635488639

iamgabrielma replied:
> I think an allowed list would be safer in case we didn't catch any events that shouldn't be decorated, also left a comment in https://github.com/woocommerce/woocommerce-ios/pull/15006#issuecomment-2635488639

Any concern regarding performance when matching events against a "safe list"? As this list would be a way bigger string matching comparison than the "exempted list". I could be overthinking it entirely, so I'll capture how long it takes when making the changes 👍

iamgabrielma replied:
I've opened a new issue to make the switch from  exempted to allowed here: https://github.com/woocommerce/woocommerce-ios/issues/15058 I think we can make the switch once we get the final event list, so we can also confirm that all is there in one go

jaclync replied:
> Any concern regarding performance when matching events against a "safe list"? As this list would be a way bigger string matching comparison than the "exempted list".

Good question about the performance impact, I'm guessing the effect is minimal if the list is under 100. On the other hand, if the POS event list does get that long, we might want to consider a different approach for the shared events (ones that could be in/outside of POS).

jaclync commented on WooCommerce/Classes/POS/Presentation/ItemListView.swift:
```diff
@@ -37,6 +37,7 @@ struct ItemListView: View {
         }
         .refreshable {
             await Task {
+                ServiceLocator.analytics.track(.pointOfSaleProductsPullToRefresh)
```
nit: this could be called at the beginning of the `refreshable` just to avoid any unforeseen issues from being run on a different thread in the Task.

iamgabrielma replied:
Good catch, updated on 9e690831f140b75aa42034463c57b25422b9e7ff

jaclync commented on WooCommerce/Classes/POS/Presentation/Item Selector/ChildItemList.swift:
```diff
@@ -34,6 +34,7 @@ struct ChildItemList: View {
         .toolbar(.hidden, for: .navigationBar)
         .refreshable {
             await Task {
+                ServiceLocator.analytics.track(.pointOfSaleVariationsPullToRefresh)
```
nit: similar to the other comment, this could be called at the beginning of the refreshable outside of the Task.

iamgabrielma replied:
Updated: 9e690831f140b75aa42034463c57b25422b9e7ff

jaclync commented on WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/PointOfSaleLoadingView.swift:
```diff
@@ -13,6 +15,25 @@ struct PointOfSaleLoadingView: View {
             .multilineTextAlignment(.center)
             Spacer()
         }
+        .onAppear {
+            trackTimeOnAppear()
+        }
+        .onDisappear {
+            trackElapsedTimeOnDisappear()
+        }
```
❓Some questions that came up: 
* Could `onAppear/onDisappear` be triggered when putting the app to the background then back, or switching to another app then back etc.?
* What if the loading view is shown outside of the initial load? From `PointOfSaleItemsController`, it sets the container state to `.loading` which shows this loading view when loading products when there are no existing products. So if there are no products in the beginning from the initial load, then triggering a reload, this view is shown again.

How about moving the tracking to `PointOfSaleItemsController`? It can be unit-tested as well to ensure the event is only tracked for the initial load (or if we want to track the event for all loads from an empty state).

iamgabrielma replied:
Answered below! https://github.com/woocommerce/woocommerce-ios/pull/15047#issuecomment-2636805047

jaclync reviewed (commented):
Thanks for tackling all the POS analytics! I took some time to catch up on the previous analytics changes during AFK. I left some comments/questions, most of the testing worked well with two issues below.

✅ 
* pos_loaded (prop milliseconds_time_elapsed_in_splash_screen)
* pos_simple_products_explanation_dialog_shown
* get_support_tapped
* pos_products_pull_to_refresh
* pos_variations_pull_to_refresh
* pos_exit_pos_menu_item_tapped
* pos_exit_pos_confirmed
* pos_checkout_tapped with `items_in_cart` prop
* pos_order_creation_success
* pos_email_receipt_tapped
* pos_email_receipt_send_tapped
* pos_create_new_order_tapped
* pos_item_removed_from_cart
* pos_back_to_cart_tapped

Issues from testing:
* If I turn off the internet after adding items to cart and then tap “Check out,” pos_order_creation_failed was tracked twice
* In the simple products dialog: when I tapped “Create an order in store management” in the dialog and the orders tab was shown, `pos_orders_add_new` got logged unexpectedly (shouldn’t have the `pos` prefix).
    * After that, I tried going back to the Menu tab but tapping on POS entry point was no-op (`hub_menu_option_tapped` with `option: pointOfSale` was logged but POS wasn’t shown). However, I couldn’t reproduce this.

Other tracking events I saw:
* pos_item_added_to_cart with `product_type` prop
* pos_card_reader_connection_success
* pos_card_present_collect_payment_canceled
* pos_card_present_collect_payment_success

Questions:
* I didn’t see an event tracked when tapping “Connect your reader.” Is this something we want to track, maybe also with the connection state?
    * Similarly, the “Connect to reader” CTA in the totals screen.
* Do we want to track the CTA tap to clear the cart?
* Do we want to track the cash payment flow? Like tapping to pay with cash and completion etc.
* Do we want to track when the success screen is shown?

iamgabrielma committed:
Merge branch 'trunk' into task/13277-pos-track-miscelaneous-events

# Conflicts:
#	WooCommerce/Classes/POS/Controllers/PointOfSaleOrderController.swift

iamgabrielma committed:
Update loaded event to use WaitingTimeTracker

iamgabrielma committed:
add orders_add_new to exemptedEvents

iamgabrielma committed:
add clear_cart_tapped event & test

iamgabrielma committed:
Add connection-disconnection events to POS side. add tests

iamgabrielma committed:
update test names

iamgabrielma committed:
move PTR events outside the task

iamgabrielma committed:
move POS order events to their own tests

iamgabrielma commented:
Thanks for the review @jaclync , this is ready for another round when you have the time. I believe I answered all items below, let me know if I missed something:

> If I turn off the internet after adding items to cart and then tap “Check out,” pos_order_creation_failed was tracked twice

I cannot replicate this one. By using the Network Link Conditioner with a 100% loss profile as well as disconnecting WIFI manually I seem to track this once only:

```
🔵 Tracked pos_order_creation_failed, properties: [error_description: URLSessionTask failed with error: The request timed out., was_ecommerce_trial: false, blog_id: -1, is_wpcom_store: false, use_gift_card: false, plan: , site_url: https://indiemelon.mystagingwebsite.com, error_context: sessionTaskFailed(error: Error Domain=NSURLErrorDomain Code=-1001 "The request timed out." UserInfo={_kCFStreamErrorCodeKey=-2102, NSUnderlyingError=0x600000d3c690 {Error Domain=kCFErrorDomainCFNetwork Code=-1001 "(null)" UserInfo={_kCFStreamErrorCodeKey=-2102, _kCFStreamErrorDomainKey=4}}, _NSURLErrorFailingURLSessionTaskErrorKey=LocalDataTask <3691169B-7429-4B45-A3D9-42F15C65CDD9>.<1>, _NSURLErrorRelatedURLSessionTaskErrorKey=(
                     "LocalDataTask <3691169B-7429-4B45-A3D9-42F15C65CDD9>.<1>"
                 ), NSLocalizedDescription=The request timed out., NSErrorFailingURLStringKey=https://indiemelon.mystagingwebsite.com/?rest_route=/wc/v3/orders, NSErrorFailingURLKey=https://indiemelon.mystagingwebsite.com/?rest_route=/wc/v3/orders, _kCFStreamErrorDomainKey=4}), store_id: c5bd46cc-1804-4f7b-badb-bb98c449127f]
```

If I switch the connection back on before the `nw_socket_handle_socket_event [C34.1.2:3] Socket SO_ERROR [60: Operation timed out]` is called, then it processes the order correctly and logs success, otherwise failed is only logged once. Does it happen often on your end?

> In the simple products dialog: when I tapped “Create an order in store management” in the dialog and the orders tab was shown, pos_orders_add_new got logged unexpectedly (shouldn’t have the pos prefix).

Good catch, added to the list here: 3f6ae935bad244048b69468a2c1ed0dbd45695e6

> After that, I tried going back to the Menu tab but tapping on POS entry point was no-op (hub_menu_option_tapped with option: pointOfSale was logged but POS wasn’t shown). However, I couldn’t reproduce this.

I couldn't neither :/

> I didn’t see an event tracked when tapping “Connect your reader.” Is this something we want to track, maybe also with the connection state? Similarly, the “Connect to reader” CTA in the totals screen.

Ahhh this one went missing! As is not tracked through the card reader service . I've added both events and tests here: a4152d021b8c9f1d679d991ef937a280dad92454

> Do we want to track the CTA tap to clear the cart?

We do, I missed this one or I was waiting for the rest of cart events. I've added it here: f4509ccb475b7ab6784b7ddf9f43c45a62dd259b

> Do we want to track the cash payment flow? Like tapping to pay with cash and completion etc.

Yes, we do, but the events are TBD for now. These are logged here: https://github.com/woocommerce/woocommerce-ios/issues/15057

> Do we want to track when the success screen is shown?

I'm not sure, the cash or card succeed might be enough. I'll ask in the P2.

> Could onAppear/onDisappear be triggered when putting the app to the background then back, or switching to another app then back etc.? What if the loading view is shown outside of the initial load? From PointOfSaleItemsController, it sets the container state to .loading which shows this loading view when loading products when there are no existing products. So if there are no products in the beginning from the initial load, then triggering a reload, this view is shown again.

That's a good question: By backgrounding the app or switching and then back I cannot see the `PointOfSaleLoadingView` re-rendering, neither when we refresh products after loading an empty list via pull-to-refresh. I'd expect so from the code in the dashboard state, as we fall into the `.loading` case but it doesn't seem to render the view? (Semi-related, if I return `.init(items: [], hasMorePages: false)` from `PointOfSaleItemsController.loadItems` I cannot pull-to-refresh, is this expected?). So we only seem to track it once on entering the POS.

Moving the event to, for example, `PointOfSaleItemsController.loadItems` seems to make it trickier since now we do call it each time we load from store. Declaring it as nil and then assigning it would help us to only call it once if we can assure that the controller is only loaded once, however, we'll have to call `.end()` each time the load products, logging events we do not want. Knowing this, do you have any recommendation about how we could track the initial load only once if it's not using appear/disappear?

jaclync commented on WooCommerce/Classes/Analytics/WooAnalyticsEvent.swift:
```diff
@@ -2486,10 +2486,12 @@ extension WooAnalyticsEvent {
 
         private enum Keys {
             static let waitingTime = "waiting_time"
+            static let milliseconds_time_elapsed = "milliseconds_time_elapsed_in_splash_screen"
```
nit: as the value is pretty specific to "splash screen", probably want to name the key accordingly. Also would be great to keep it camel case.

iamgabrielma replied:
True that, updated: https://github.com/woocommerce/woocommerce-ios/pull/15047/commits/47924eac8e1d61894ada8234adadc5b42db9e48b

jaclync commented on WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/PointOfSaleLoadingView.swift:
```diff
@@ -21,6 +41,9 @@ private extension PointOfSaleLoadingView {
         static let progressViewSpacing: CGFloat = 72
     }
+    enum Constants {
+        static let toMilliseconds: Double = 1000.0
```
nit: this doesn't seem needed anymore.

iamgabrielma replied:
Updated! https://github.com/woocommerce/woocommerce-ios/pull/15047/commits/9c57d76ed72a09739a06b811abbcbd9af6ef60b0

jaclync commented on WooCommerce/WooCommerceTests/POS/Controllers/PointOfSaleOrderControllerTests.swift:
```diff
@@ -235,6 +238,57 @@ struct PointOfSaleOrderControllerTests {
         #expect(mockPaymentCelebration.celebrationWasCalled == true)
     }
+
+    struct AnalyticsTests {
```
Nice thanks for adding these tests!

jaclync commented on WooCommerce/Classes/Analytics/WooAnalyticsEvent.swift:
```diff
@@ -2504,6 +2506,8 @@ extension WooAnalyticsEvent {
             case .appStartup:
                 return WooAnalyticsEvent(statName: .applicationOpenedWaitingTimeLoaded, properties: [Keys.waitingTime: elapsedTime])
+            case .pointOfSaleLoaded:
+                return WooAnalyticsEvent(statName: .pointOfSaleLoaded, properties: [Keys.milliseconds_time_elapsed: elapsedTime])
```
❕ During testing it looks like the prop is recording the time in seconds, whereas it's expected to be milliseconds:

```🔵 Tracked pos_pos_loaded, properties: [milliseconds_time_elapsed_in_splash_screen: 2.351064920425415, ...]```

iamgabrielma replied:
I think that's not the case, is it? I might be reading this wrongly but when we log the event on [WaitingTimeTracker ](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/Analytics/WaitingTimeTracker.swift#L27-L30) we log the elapsed time, so in this case we're still capturing it in milliseconds right? So 1.9ms in this instance, which we pass to the event:
```
po currentTimeInMillis()
> 1738832706.298878

po waitingStartedTimestamp
> 1738832690.2819052

po elapsedTime (as currentTimeInMillis() - waitingStartedTimestamp)
> 1.9627389907836914
```

iamgabrielma replied:
Actually the calculation seems off too 1738832706.298878 - 1738832690.2819052 it's not what we get but ~16, checking as I have just used the WaitingTimeTracker that was already there 👀

iamgabrielma replied:
In https://github.com/woocommerce/woocommerce-ios/pull/15047/commits/4d4986192481032e61a5c824e7e7e168b32a1920 I've updated the interface of WaitingTimeTracker as there was a very miss-leading property in the initializer: `currentTimeInMillis` captures the time in seconds, not in milliseconds.

I've added a new `endInMilliseconds()` method just for our use case, so I avoid touching the current implementation much, and now captures the elapsed time properly:

```
🔵 Tracked pos_pos_loaded, properties: [is_wpcom_store: false, was_ecommerce_trial: false, milliseconds_time_elapsed_in_splash_screen: 2172.056198120117, site_url: https://indiemelon.mystagingwebsite.com, plan: , store_id: c5bd46cc-1804-4f7b-badb-bb98c449127f, blog_id: -1]
```

jaclync commented:
> > In the simple products dialog: when I tapped “Create an order in store management” in the dialog and the orders tab was shown, pos_orders_add_new got logged unexpectedly (shouldn’t have the pos prefix).
> 
> Good catch, added to the list here: [3f6ae93](https://github.com/woocommerce/woocommerce-ios/commit/3f6ae935bad244048b69468a2c1ed0dbd45695e6)
> 
> > After that, I tried going back to the Menu tab but tapping on POS entry point was no-op (hub_menu_option_tapped with option: pointOfSale was logged but POS wasn’t shown). However, I couldn’t reproduce this.
> 
> I couldn't neither :/

Actually, I just encountered this in the first run of PR testing again today. I did some debugging, but couldn't figure out the cause in the time I had: the `fullScreenCover` content was called to display the POS entry point view, but the actual view was hidden and the navigation stack in the view hierarchy was grayed out:

<img width="2032" alt="Screenshot 2025-02-06 at 2 46 22 PM" src="https://github.com/user-attachments/assets/078200fb-445f-4fc2-8ccb-790de0420b44" />

I also recalled the order creation form not being shown in the orders tab. It might be some UI issue related to race conditions.

jaclync commented:
> > If I turn off the internet after adding items to cart and then tap “Check out,” pos_order_creation_failed was tracked twice
> 
> I cannot replicate this one. By using the Network Link Conditioner with a 100% loss profile as well as disconnecting WIFI manually I seem to track this once only:
> 
> ```
> 🔵 Tracked pos_order_creation_failed, properties: [error_description: URLSessionTask failed with error: The request timed out., was_ecommerce_trial: false, blog_id: -1, is_wpcom_store: false, use_gift_card: false, plan: , site_url: https://indiemelon.mystagingwebsite.com, error_context: sessionTaskFailed(error: Error Domain=NSURLErrorDomain Code=-1001 "The request timed out." UserInfo={_kCFStreamErrorCodeKey=-2102, NSUnderlyingError=0x600000d3c690 {Error Domain=kCFErrorDomainCFNetwork Code=-1001 "(null)" UserInfo={_kCFStreamErrorCodeKey=-2102, _kCFStreamErrorDomainKey=4}}, _NSURLErrorFailingURLSessionTaskErrorKey=LocalDataTask <3691169B-7429-4B45-A3D9-42F15C65CDD9>.<1>, _NSURLErrorRelatedURLSessionTaskErrorKey=(
>                      "LocalDataTask <3691169B-7429-4B45-A3D9-42F15C65CDD9>.<1>"
>                  ), NSLocalizedDescription=The request timed out., NSErrorFailingURLStringKey=https://indiemelon.mystagingwebsite.com/?rest_route=/wc/v3/orders, NSErrorFailingURLKey=https://indiemelon.mystagingwebsite.com/?rest_route=/wc/v3/orders, _kCFStreamErrorDomainKey=4}), store_id: c5bd46cc-1804-4f7b-badb-bb98c449127f]
> ```
> 
> If I switch the connection back on before the `nw_socket_handle_socket_event [C34.1.2:3] Socket SO_ERROR [60: Operation timed out]` is called, then it processes the order correctly and logs success, otherwise failed is only logged once. Does it happen often on your end?

Ah never mind, the multiple `pos_order_creation_failed` events were from retrying while the device was still offline.

jaclync commented:
> > Could onAppear/onDisappear be triggered when putting the app to the background then back, or switching to another app then back etc.? What if the loading view is shown outside of the initial load? From PointOfSaleItemsController, it sets the container state to .loading which shows this loading view when loading products when there are no existing products. So if there are no products in the beginning from the initial load, then triggering a reload, this view is shown again.
> 
> That's a good question: By backgrounding the app or switching and then back I cannot see the `PointOfSaleLoadingView` re-rendering, neither when we refresh products after loading an empty list via pull-to-refresh. I'd expect so from the code in the dashboard state, as we fall into the `.loading` case but it doesn't seem to render the view? (Semi-related, if I return `.init(items: [], hasMorePages: false)` from `PointOfSaleItemsController.loadItems` I cannot pull-to-refresh, is this expected?). So we only seem to track it once on entering the POS.

Maybe it will never reach the state to show this loading view, it looks like when there are no eligible products this view is shown:

<img src="https://github.com/user-attachments/assets/79193de1-0a3c-448e-916a-65428a5569f0" width="500" />

Is this the view you mentioned there is no PTR?

In any case, if the UI logic changes in the future to show this loading view, the event might be unintentionally tracked.
 
> Moving the event to, for example, `PointOfSaleItemsController.loadItems` seems to make it trickier since now we do call it each time we load from store. Declaring it as nil and then assigning it would help us to only call it once if we can assure that the controller is only loaded once, however, we'll have to call `.end()` each time the load products, logging events we do not want. Knowing this, do you have any recommendation about how we could track the initial load only once if it's not using appear/disappear?

I was thinking about adding a private boolean variable to track whether the initial load event has been tracked + unit tests to ensure only the initial load is tracked. Any other cleaner solutions would be great too.

jaclync reviewed (approved):
Thanks for the updates! LGTM once the ❕ comment is addressed.

iamgabrielma committed:
Merge branch 'trunk' into task/13277-pos-track-miscelaneous-events

# Conflicts:
#	WooCommerce/WooCommerceTests/POS/Models/PointOfSaleAggregateModelTests.swift

iamgabrielma committed:
Update to iOS17 min requirement

iamgabrielma committed:
remove unused const

iamgabrielma committed:
rename event key

iamgabrielma commented:
Thanks again for the review!  I just have a question regarding the milliseconds tracking [here](https://github.com/woocommerce/woocommerce-ios/pull/15047#discussion_r1944361739).

> Is this the view you mentioned there is no PTR?
> In any case, if the UI logic changes in the future to show this loading view, the event might be unintentionally tracked.
> I was thinking about adding a private boolean variable to track whether the initial load event has been tracked + unit tests to ensure only the initial load is tracked. Any other cleaner solutions would be great too.

Yes, that's the one. It also seems we're removing the `.loading` state from PTR as we move to Observable, so I'll hold this change a bit so it doesn't get too messy. I'll handle that event separately to be sure it's only called on POS start.

iamgabrielma committed:
Update WaitingTimeTracker interface. capture elapsed time in ms

iamgabrielma committed:
update WaitingTimeTracker interface in test target
